src/rdb.c

```diff
@@ -937,39 +943,73 @@ ssize_t rdbSaveObject(rio *rdb, robj *o, robj *key, int dbid) {
         }
     } else if (o->type == OBJ_HASH) {
         /* Save a hash value */
-        if (o->encoding == OBJ_ENCODING_LISTPACK) {
-            size_t l = lpBytes((unsigned char*)o->ptr);
+        if ((o->encoding == OBJ_ENCODING_LISTPACK) ||
+            (o->encoding == OBJ_ENCODING_LISTPACK_EX))
+        {
+            unsigned char *lp_ptr = hashTypeListpackGetLp(o);
+            size_t l = lpBytes(lp_ptr);
 
-            if ((n = rdbSaveRawString(rdb,o->ptr,l)) == -1) return -1;
+            if ((n = rdbSaveRawString(rdb,lp_ptr,l)) == -1) return -1;
             nwritten += n;
         } else if (o->encoding == OBJ_ENCODING_HT) {
             dictIterator *di = dictGetIterator(o->ptr);
             dictEntry *de;
+            int with_ttl = 0; /* whether there is at least one field with a valid TTL */
+            uint64_t ttl = 0;
 
+            /* save number of fields in hash */
             if ((n = rdbSaveLen(rdb,dictSize((dict*)o->ptr))) == -1) {
                 dictReleaseIterator(di);
                 return -1;
             }
             nwritten += n;
 
+            /* use standard hash format if no TTL was set for any field */
+            if (hashTypeGetMinExpire(o) != EB_EXPIRE_TIME_INVALID)
+                with_ttl = 1;
+
+            /* save all hash fields */
             while((de = dictNext(di)) != NULL) {
                 hfield field = dictGetKey(de);
                 sds value = dictGetVal(de);
 
+                /* save the key */
                 if ((n = rdbSaveRawString(rdb,(unsigned char*)field,
                         hfieldlen(field))) == -1)
                 {
                     dictReleaseIterator(di);
                     return -1;
                 }
                 nwritten += n;
+
+                /* if this is a hash table with a TTL for at least one field,
+                 * save it in the "with metadata" format - one byte for included
+                 * metadata types, and than the types themselves (note that the
+                 * order is fixed, i.e. future extensions may only add new
+                 * metadata types at the end) */
+                if (with_ttl && hfieldIsExpireAttached(field)) {
+                    ttl = hfieldGetExpireTime(field);
+                } else {
+                    ttl = 0;
+                }
+
+                /* save the value */
                 if ((n = rdbSaveRawString(rdb,(unsigned char*)value,
                         sdslen(value))) == -1)
                 {
                     dictReleaseIterator(di);
                     return -1;
                 }
                 nwritten += n;
+
+                /* save the TTL */
+                if (with_ttl) {
+                    if ((n = rdbSaveLen(rdb, ttl)) == -1) {
+                        dictReleaseIterator(di);
+                        return -1;
+                    }
+                    nwritten += n;
+                }
```

Comment:
Merged, but remove the internal condition outside (this condition is not avoided regardless of its location, but it's simpler to understand when it's not used as a function param IMHO).
The original location was actually to allow saving the "options" filed (whether value and/or TTL exists) but since this was deleted, it was indeed wrongly placed.